python/pyarrow/__lib_pxi/array.pyi

```diff
@@ -0,0 +1,4294 @@
+# Licensed to the Apache Software Foundation (ASF) under one
+# or more contributor license agreements.  See the NOTICE file
+# distributed with this work for additional information
+# regarding copyright ownership.  The ASF licenses this file
+# to you under the Apache License, Version 2.0 (the
+# "License"); you may not use this file except in compliance
+# with the License.  You may obtain a copy of the License at
+#
+#   http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing,
+# software distributed under the License is distributed on an
+# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+# KIND, either express or implied.  See the License for the
+# specific language governing permissions and limitations
+# under the License.
+
+import datetime as dt
+import sys
+
+from collections.abc import Callable
+from decimal import Decimal
+
+if sys.version_info >= (3, 11):
+    from typing import Self
+else:
+    from typing_extensions import Self
+from typing import (
+    Any,
+    Generic,
+    Iterable,
+    Iterator,
+    Literal,
+    LiteralString,
+    TypeVar,
+    overload,
+)
+
+import numpy as np
+import pandas as pd
+
+from pandas.core.dtypes.base import ExtensionDtype
+from pyarrow._compute import CastOptions
+from pyarrow._stubs_typing import (
+    ArrayLike,
+    Indices,
+    Mask,
+    Order,
+    SupportArrowArray,
+    SupportArrowDeviceArray,
+)
+from pyarrow.lib import (
+    Buffer,
+    Device,
+    MemoryManager,
+    MemoryPool,
+    MonthDayNano,
+    Tensor,
+    _Weakrefable,
+)
+from typing_extensions import deprecated
+
+from . import scalar, types
+from .device import DeviceAllocationType
+from .scalar import NullableCollection, Scalar
+from .types import (
+    DataType,
+    Field,
+    MapType,
+    ListType,
+    _AsPyType,
+    _BasicDataType,
+    _BasicValueT,
+    _DataTypeT,
+    _IndexT,
+    _RunEndType,
+    _Size,
+)
+
+@overload
+def array(
+    values: NullableCollection[bool],
+    type: None = None,
+    mask: Mask | None = None,
+    size: int | None = None,
+    from_pandas: bool | None = None,
+    safe: bool = True,
+    memory_pool: MemoryPool | None = None,
+) -> BooleanArray: ...
+@overload
+def array(
```

Comment:
Is this approach scalable? I notice this doesn't seem to include some extension types such as tensor arrays?

Reply:
Are you asking about annotating constructors per type? Overload approach is nice because it expresses the relationship between the argument and return types. We can use a single parameter type union and a single return union, but we then lose the relationship.

`FixedShapeTensorArray` is on line 4578, I'm not sure about other extension types. If we use type checkers to validate our tests we should get errors if our symbols are not annotated or if annotations disagree with tests. Our test coverage is pretty good and we can assume every symbol is tested at least once.